cpp/src/join/semi_join.cu

```diff
@@ -60,26 +61,84 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
     return result;
   }
 
-  // Materialize a `flagged` boolean array to generate a gather map.
-  // Previously, the gather map was generated directly without this array but by calling to
-  // `map.contains` inside the `thrust::copy_if` kernel. However, that led to increasing register
-  // usage and reducing performance, as reported here: https://github.com/rapidsai/cudf/pull/10511.
-  auto const flagged =
-    cudf::detail::contains(right_keys, left_keys, compare_nulls, nan_equality::ALL_EQUAL, stream);
```

Comment:
Does this issue affect other uses of `cudf::detail::contains`? Should we be changing that function's implementation instead of just the semi-join implementation? (I haven't formed an opinion on this question yet, need to read more code first.)

Reply:
Right question. Changing the implementation will completely fix this, but requires new FEA from cuco, which is under way: https://github.com/NVIDIA/cuCollections/pull/191

Reply:
This concerns me as well. It seems an issue with lots of duplicate keys in general and we just found this instance to be a problem so far.

Reply:
Currently, the only use case of `cudf::detail::contains` is in lists operations.

Reply:
Sound reasonable. Then I'll address that concern in another separate PR. I would like to keep `cudf::detail::contains` separated from semi-anti-join for 22.08 to prevent any last-minute surprising performance issue.

Reply:
We can ask Jake for a quick review since the change is basically the same as https://github.com/NVIDIA/cuCollections/pull/175. The issue must affect other use cases of `detail::contains` but just not unveiled by the existing benchmarks yet.

Reply:
That's great. If we have #cuco/191 merged quickly then I can have a complete fix up for `detail::contains` without a temp fix.

Reply:
Working on it now, should be ready very quickly.